### ISSUE TITLE
What about the people behind socketstream project

### ISSUE THREAD

**Issue by haohello** [socketstream/socketstream-0.4#17]
I really like the idea of the modular approach for the socketstream-0.4, even though the future for the project sounds promising, however the development progress will surely block most of the people who are interested in this project. 

As far as I could feel, this project seems to be backed by only one person, and it will be extremely difficult for only one person to accomplish the goal for this project especially in nowadays. I'm surprised to find out that how the projects like [jQuery](https://jquery.org/), [Linux](http://www.linuxfoundation.org/) and also [openstack](http://www.openstack.org/foundation/) advance in this rapidly changing environment. They are strong enough with their organization behind the projects and bring the confidence to everyone involved, and they're still evolving on the way.

I know there are surely a lot guys out there who are willing to contribute in some way to this project, and I think they need to be inspired and always need the value of existence while involved in this project.

For a toy project of only one person, I can see that you were very passionate in the very begining and this kind of passion only last for not too long and the project progress and commit frequency slowed down gradually. 

 In the long run, I think you need to bring more people in this project and always keep them inspired.

 As a leading developer for this project, you've already had a lot followers here,  the other thing to think about is how to keep them more involved and organized in a way as a team to keep this project move forward.  http://changingminds.org/disciplines/leadership/articles/manager_leader.htm

Please let me know how you're thinking.

**Comment by paulbjensen** [comment]
I'd like to offer some thoughts.

In my opinion, SocketStream needs better documentation and more tutorials on how to accomplish various tasks in the developer's workflow (e.g. building app features, doing testing, deployment etc). 

There are enough developers using SocketStream who have experience in doing these things, but there isn't a single place where all of this knowledge can be contributed and shared.

I've wanted to help out with this, but my work-life-balance has meant that time doesn't exist for doing work on side projects, including this. I'm hoping to manage my time better so that I can.

With 0.3, there have been quite a few pull requests and npm modules built around that version of SocketStream. I wish that those code contributions could have lived on in the 0.4 codebase, but it's a small price to pay for allowing the framework to be rewritten so that it can be better. 

My hope is that in the future, these gaps around SocketStream get filled, that the community will grow, and with that, an ecosystem will emerge.

**Comment by arxpoetica** [comment]
I can't really speak to what @owenb would want in terms of direct repo contributors, but I do think that needs to happen at some point. I've been heavily invested in SS, but if I'm completely honest, Meteor.js is becoming more and more attractive as it grows. Yet I would rather stay with something like SocketStream. It's just a more attractive offering for multiple reasons.

I do agree with the sentiment from @haohello from above. I feel concerned that the project isn't quite open enough. I also agree with @paulbjensen about a need for a better marketing website with really solid documentation. I'm happy and willing to work on that in pieces here and there -- as I said, I'm invested. But I don't think documentation is enough. I'm not exactly certain what needs to happen, but code contribution needs to not only be encouraged, but it needs to happen.

Some suggestsions toward that end -- and perhaps this is needed _more_ than better documentation on how SS works -- is to write up _really good_ documentation about _how_ to get involved. If so many of us (myself included) feel a need to get involved, but don't know 100% how to, that's a good indication that we need to work up a working system.

It doesn't have to be complicated or overly involved -- simpler is even better. But people need to know exactly how they can do that with the confidence that good code/pull requests will make it into the project.

I'm ready to start if @owenb can help lead the way. I actually created a Trello board once to this end. Maybe it's worth kicking that thing back around. OR maybe we just use GitHub. (The nice thing about Trello, is anyone can be invited without worrying about commit rights.) Here's the org. page https://trello.com/socketstream and here's the org. board page https://trello.com/board/socketstream-development/5112e449bf9b88457d00d040 I've made it public, but I think you'll have to request to be a member. If you can't request there, just request here, and I'll try and add you.

**Comment by owenb** [comment]
Thanks very much for your thoughts. I've been thinking about this subject a lot recently, especially since receiving @haohello message.

Unfortunately I don't have the free time to work on SocketStream at the moment, so I'd love to find ways for other people to contribute. Not just minor code changes, but real development work towards the goals of 0.4.

The good news is that the 0.4 code is now in a pretty good state and most of the hard design decisions have been made. But there is still much to be done.

My ideal is to find a few people who are interested and motivated to take things forward. We can test the water with a few pull requests, but I'm very keen to trust people with full commit rights and ownership over certain areas.

Of course, I'll always be happy to discuss ideas, review code and provide direction when required.

Naturally I'm looking for good coders, but what's far more important is that core contributors agree with the overall vision of the project. 

In particular, I have two major goals for SocketStream 0.4:

#### 1. More toolkit, less framework

One of the best things I've ever read on the Internet is Substack's blog post called [Many Things](http://substack.net/many_things). He neatly sums up many of the issues I have been feeling recently about using frameworks vs libraries.

I'm keen to see SocketStream progress more as a set of libraries that do one thing well (optionally combined into a toolkit) rather than a big monolithic framework.

I really like James' ideas to create more recipes and community, rather than abstractions which remove key decisions away from the developer. It's very important to me to continue nudging SocketStream 0.4 in this direction.

#### 2. Simple over Easy

We all love things which are easy. SocketStream 0.1 proved to be a huge hit as everyone could download it and get up and running with a realtime website in minutes. Two years ago, this was a huge deal!

A few months ago I saw Rich Hickey's utterly amazing talk [Simple vs Easy](http://www.infoq.com/presentations/Simple-Made-Easy). Ever since then I've not stopped thinking about how to build simpler software. I now realise that Simple and Easy are often two completely separate things.

Simple often means more code, more modules, more boilerplate, and more to learn up front. But ultimately it rewards you with an app that's easier to extend and maintain in the future. If you haven't watched that video, go watch it now. It will change the way you think about software forever.

---

So these are the two guiding principles I want others to follow when contributing to SocketStream. If we can relentlessly focus on building simple things, it will become a great product - far superior in many ways to everything that's out there at the moment.

That's not to say it can't be easy to use too. Good documentation will certainly help there! But it must be made simple before it can be made easy.

Please let me know your thoughts, guys.

If you're interested in becoming a core contributor, please get in touch below or privately, if you prefer, and let's get the ball rolling.

Owen

**Comment by wamatt** [comment]
I really love the idea of what SocketStream can become, however also agree with the sentiment expressed here. It needs to be marketed better and have more contributors to prosper longterm.

**Comment by bloodyKnuckles** [comment]
The thought comes to mind, does socketstream have to have a lot of people using it to be a great resource?

**Comment by wamatt** [comment]
> The thought comes to mind, does socketstream have to have a lot of people using it to be a great resource?

Maybe, maybe not. That really is a tangential issue since you are talking about users. What it clearly does need however, is a core group of passionate _contributors_, for long term success. Without that, the users look elsewhere simply because they do not want to build projects on top of a framework that will be abandoned. 

Abandoned or dying projects can quickly become outdated: bugs remain unfixed etc. Momentum is a key part of the opensource lifecycle.

**Comment by davisford** [comment]
> Abandoned or dying projects can quickly become outdated: bugs remain unfixed etc. Momentum is a key part of the opensource lifecycle.

This is so true -- I'm sorry to hear that @owenb is lacking the time to commit here.  You've started something really great, and I also don't want to abandon it b/c I've also invested quite a lot in it -- having built several apps with ss now.  I'm happy to also help contribute -- I would like ss to continue to evolve and thrive.  I do share the same goals you stated, Owen, but like @americanyak I'm interested in finding a good entry point, so let me know how I can assist.

**Comment by paulbjensen** [comment]
It would be great to get more detail on what remains to be done for 0.4; a TODO list or a set of trello tasks or a list of labelled Github issues.

I'd be tempted to recommend Github Issues; you can specify feature requirements, have someone assign themselves to the issue, submit a pull request, and merge it in once it's good. One place to track everything.

**Comment by mdlawson** [comment]
I'd be very interested in contributing as well, but like others have said, it would be great to see some github issues outlining what needs to be done. Even just a general overview of the projects immediate goals moving forward would be great.

**Comment by arxpoetica** [comment]
@owenb I'm in. I'll do whatever I can to give good contributions, including code review, etc. etc. I think I'm familiar enough with the product to know where I can add the most value. I'd also be happy to lead the way on documentation. In fact, if you can give me commit rights, I'll focus on that up front. As for Trello, since you're opening this up, there's no point in using something else. Let's use GitHub. My recommendation, as @paulbjensen suggested is to start with a list of TODO and FIXME issues in the GitHub account, start sorting them. That and focus on how to get people set up. Part of a working open source community is not just commit rights, but also paying attention to pull requests. So I'd like to see documentation advising people on how to fork, setup local, and do pull requests for development on SS.

In fact, I'm going to need that. I'll open a ticket for that request alone. :)

**Comment by arxpoetica** [comment]
https://github.com/socketstream/socketstream-0.4/issues/19 Lets discuss documentation and SS dev setup needs over here. That way this ticket can continue to focus on the bigger picture of involving the open source community in the development of SocketStream.

**Comment by owenb** [comment]
Hi guys

In the spirit of [Felix's Github hack](http://felixge.de/2013/03/11/the-pull-request-hack.html) I've added @americanyak and @davisford to this 0.4 project and Prism - the heart of SocketStream 0.4. Guys, you've been big supporters of SocketStream for a long time - thank you!

My good friend @paulbjensen has been involved in SocketStream from day one and can also commit to everything, including the main repo. If you need anything changing on https://github.com/socketstream/socketstream and I'm not around to help, please ask Paul (though I know he's also a busy chap these days).

When I get a moment I will go through the existing 0.4 code and outline what I think needs to be done to finalise the design. Once we have the design right, we can prioritise the docs. Only then can should we think about marketing things hard.

Let the big experiment begin :)

**Comment by davisford** [comment]
Thanks @owneb -- my free time did just open up a little bit too, so I should be able to help out more.  Looking forward to it.

**Comment by haohello** [comment]
Sorry to bring up this serious topic, my primary goal of starting this topic was to see if @owenb and other guys are still confident with the project' long term direction.

The other important aspect that I was referring, the people or organization behind the project, for a project to succeed it usually requires a lot of collective efforts, rather than one man to guide the directions, that's why i'm emphasizing the organization, building an organization with certain rules to guide the collaborations. 

From the goals that @owenb stated in the README file, socketstream project wants to be an alternative to meteor.js, however for the current situation of socketstream project, it'll be impossible for us as guerrillas to beat them. 

Guerrilla and regular army are two completely different systems, statistically speaking regular army will always win in the competition.

**Comment by arxpoetica** [comment]
I don't really view it as a competition. It's just an alternative that many of us would rather go with.

**Comment by owenb** [comment]
It's not a competition.

Geoff and Matt (the main Meteor guys) are super-intelligent, lovely people. I have met them a number of times and have a great deal of respect for them and their crazy-ambitious endeavour. They fully deserve to succeed and I'm certain they will.

But Meteor is not so much a framework as an entire end-to-end platform and ecosystem. There is no way we can, or should, want to emulate or compete with this.

SocketStream has always been about finding the best way to build single-page realtime apps using Node.js. That's the scope - no more.

Today, more than ever, this needs to be done in a way which uses and integrates well with other great libraries such as Express, rather than pretending they don't exist - or trying to reinvent everything from scratch.

If this is done right SocketStream 0.4 will have a lot less code than 0.3 and yet offer infinitely more choice and flexibility.

**Comment by wamatt** [comment]
@owenb not sure if you've chatted to the TowerJS guys? They've recently had a major change in direction for 0.5. They asked me not to post on HackerNews and trying to keep it hush, but it seems they have a super ambitious goals.

They are completely rewriting in JS (it was CS before) and going heavily component based. This struck me similar to the ideals you had for the vision of this project originally. It's a bit far out, but what about combining forces?

Also I agree from some aspects this project is not competing with Geoff and Matt (they are cool people). However, if we consider the more abstract case: there exists are a non-infinite amount of users, therefore there necessarily must be competition for mindshare of those resources, if we assume people do not code their projects in both Meteor and SS. 

So from that aspect it's useful to think about competition and the big picture, rather than competition between people or their feelings.

**Comment by arxpoetica** [comment]
@owenb that last comment is worthy of being in the documentation. :) Just sayin'. (Your comment, not @wamatt, though it's a good comment too, ha.)

**Comment by RomanMinkin** [comment]
Hi guys,

Today I was at <a href="http://www.10gen.com/events/mongonyc-2013">MongoNYC 2013</a> and I met Diane M from <a href="http://www.redhat.com/">redhat.com</a>. 

First of all she is an very nice person and open source fan =), but at conference she was representing https://www.openshift.com/, which is new hosting solution from RedHat. I told here about SocketStream project, and that project needs hosting for demo and documentation and some community support as well. So Diane liked the idea of the project and she said that she can help with providing a hosting space at OpenShift for SocketStream for free and also can help with promoting SocketStream in RedHat community (or maybe somewhere else too, which is even better), I mean somebody in communities could be interested in using SS and maybe contribution.

OpenSift does support WebSockets https://www.openshift.com/blogs/paas-websockets through `node-proxy` ( I made some research ;) ).
And it's also have a free plan https://www.openshift.com/products/pricing, but i think it's even better to have on board person like Diane who interested in helping. 

I just like SS project and I'm really interested in it, so i'm just trying to help out =)

So let me know what do you think.

**Comment by arxpoetica** [comment]
That sounds great!

**Comment by RomanMinkin** [comment]
Hi everyone,

Well, i played with OpenShift today, unfortunately it has some problems with managing node.js versions. It require additional pre/post build code etc. So i spent almost one day and my app still doesn't work the way i expected. And it's slow on application update. But still free =)

@americanyak , I remember you <a href="https://github.com/socketstream/socketstream-0.4/issues/19#issuecomment-18648415" >made a request</a> to http://opensource.nodejitsu.com/, did they respond?

**Comment by haohello** [comment]
hi guys, just came across a [node.js real time game server framework based on socket.io](https://github.com/NetEase/pomelo) backed by a Chinese online game operator who has been traded in Nasdaq since 14 years ago, really a worth to take a look at, I'm thinking about trying it out in my other apps.

**Comment by paulbjensen** [comment]
@haohello Thanks for that, I've heard of it before but I haven't taken a close look. Interesting to see how they run the socket management and backend logic under separate processes.

@RomanMinkin I guess OpenShift is relatively new and so needs to iron out teething issues, still awesome that Diane offered support. I remember Owen tried Nodejitsu a while back and said it worked really well. 

The web site would have to be hosted on one platform or the other (or even somewhere else), but there's no reason why we couldn't partner with both to show how to deploy SocketStream apps on each respective platform, as well as promote existing solutions to the community (like Dashku, shameful plug).

## tl;dr

1 - Time management -  If you find you have less free time, time manage more effectively; make use of those unexpected pockets of free time that come about.
2 - Clear-to-neutral - Do things the night before, so that you don't have to do them the next day. Make decisions about what you want to do with your free time, before that free time arrives.

## Time management

On a separate note, I've been thinking about how to deal with the fact that like others on this project, my time on working on stuff like SocketStream is currently constrained.

What I've come to realise is that time management is the key. 6 months ago I had lots of spare time on my hands, but since then work had to deliver 2 projects in a very short period of time, and I met a girl. There isn't much spare time.

I get unexpected pockets of free time, say waking up at 6am on a Saturday. In that free time, I could make use of myself and do something.

## Clear to neutral

I tried to explain it, but this article I googled does a much better job: http://www.asianefficiency.com/habits/clearing-to-neutral/

## Summary

If I can combine the philosophy of clear-to-neutral with the pockets of free time I get, I believe I can do more.

**Comment by RomanMinkin** [comment]
@paulbjensen I agree. BTW I have running test app at OpenShift, going to make an additional branch for `SS 0.3` with seted up demo.

**Comment by arxpoetica** [comment]
Okay. Making some noise again. I'm back on an SS project, so ready to roll on working. @RomanMinkin any more word from OpenShift? I want to start in on some basic documentation.

**Comment by RomanMinkin** [comment]
@americanyak , nothing yet, which is wired, but I was able to run test app on OpenShift. So the question is how to make SS demo? Should it be different Github project for this or not. The one thing with i do not like is that you have to use environment variables like 

``` javascript
process.env.OPENSHIFT_NODEJS_I
process.env.OPENSHIFT_INTERNAL_IP
```

To run the HTTP and WebSocket servers. 
So i can help with setting up demo app and it would be great if @owenb would be able to set up an account there https://www.openshift.com/

**Comment by davisford** [comment]
@americanyak glad to hear it -- i just want to say i thought i could commit a little more but i'm crazy.  i have two jobs (M-F) and (S-S) and two kids under 5 right now.  i'm nuts.  in the short term, i'll partake in any issues, pull requests, etc., but right now i'm burning the candle @ both ends so i'm not a good candidate for taking the torch.  hope to change that soon.

**Comment by owenb** [comment]
Hey guys

Thanks for helping out.

So, I've written up the first 0.4 task that needs to be done. It's Issue #21.

If anyone would like to take this, let me know (just comment on the issue) and we can discuss it further. I'd love to get others involved in the core development. I'm sure we'll get 0.4 done in no time if we can get a steady stream of contributions going. I'll write up some more Issues soon.

I've also merged a pull request from @paulbjensen in today and aim to get another 0.3 release out soon, complete with updated dependencies.

Also, I see there have been a few interesting projects using SocketStream over the last few weeks:

VMUX - WebRTC - http://vmux.co 
Bitjoy - Realtime Bitcoin data, powered by @socketstream bitjoy.org
Flying drones and controlling lasers! vine.co/v/hK700EXhH7Q

I'm not publicising these at the moment, but I'll make sure these examples are on our main website when 0.4 is ready.

@RomanMinkin I'd like to keep the main www.socketstream.org site on EC2 as it is now, but it would be great to have some demo apps on the OpenShift stack. I'll look into it more and see about creating an account. Thanks for your help here.

---

Just so you all know, work-wise my primary focus these days is developing my own 'startup' app. I've had a big idea for many years (prior to SocketStream) and am finally getting time to make it a reality. I'm bootstrapping development work with consulting and contracting which is working out well.

This doesn't leave me with a huge amount of time to spare for SocketStream right at the moment, but as so many of you are passionate about the project, I'm more than happy to devote chunks of time every week to help out with 0.3 issues and guide 0.4 development.

**Comment by arxpoetica** [comment]
@owenb hit me up when you have a moment. I want to discuss wayz going forward. I think there are still some gaps that can be filled that will help the momentum of this project; as well, I want to discuss my own involvement. I'll try and hang around the IRC.

**Comment by arxpoetica** [comment]
(And totally understand, @davisford)

**Comment by paulbjensen** [comment]
Hi,

Some of you will have seen last week's announcement, but it's worth reposting here. https://t.co/m7FGsDS8J0

In short, Owen has given me the role of managing the project. This frees up Owen to pursue the startup idea he's been wanting to do for a long time, and allows SocketStream to resume active development.

On that note, we're currently porting the source code from CoffeeScript to Javascript (see here: https://groups.google.com/forum/?fromgroups#!topic/socketstream/jhomwoJBMDo ). If anyone wants to dig in, take a CoffeeScript file in the src folder, add the comments from it into it's compiled version in the lib folder, and lint it with JSHint, then PR it into the feature/one-language branch. I've done quite a few of them, but there's a lot more to do.

Robert aka @americanyak is heading up documentation.    

@RomanMinkin Fostering a partnership with Red Hat would be great. Could you get in touch with Diane and let her know we'd like to build a demo running on OpenShift, write it up as a blog post, and promote it?

@haohello It would be good to get your thoughts on this recent change.

**Comment by RomanMinkin** [comment]
Hi @paulbjensen,

Sure, i would love do that =)

Would you mind to register an official account at https://www.openshift.com/ ? And if give me an access I can set everything up, probably we need an additional sub-branch or even better a dedicated sub-project for the demo apps including OpenShift.

I sow Owen's letter, I was busy a bit last week. Well it's sad from one side, but from the other I hope that SS community is strong enough keep this project running. I would love to do all what i can for that.

**Comment by arxpoetica** [comment]
I won't pretend to speak for @owenb, so I'll only pretend to pretend to speak for him, ha. I think he still will be involved peripherally, as it is still his baby. I imagine any movement will still be satisfactory to him, as well (and here I speak for myself), I aspire for his blessing on directions we take things. Now I'll step back and let him speak for himself, ha ha. But I think it was a really good move on his part.

**Comment by owenb** [comment]
Thanks @americanyak 
That's exactly what I think :)

**Comment by paulbjensen** [comment]
Hi @haohello, has your question been answered?
